File: jlox-rust/src/interpreter.rs
use crate::expr;
use expr::{Expr, Value};
use crate::stmt;
use crate::stmt::Stmt;
use crate::token::Token;
use crate::token_type::TokenType;
use crate::environment::Environment;
use std::rc::Rc;
use std::cell::RefCell;

pub struct Interpreter {
    environment: Rc<RefCell<Environment>>
}

impl Interpreter {
    pub fn new() -> Interpreter {
	Interpreter {
	    environment: Rc::new(RefCell::new(Environment::new()))
	}
    }

    pub fn interpret(&mut self, stmts: Vec<Stmt>) {
	for stmt in stmts {
	    self.execute(&stmt);
	}
    }

    fn execute(&mut self, stmt: &Stmt) {
	stmt.accept(self);
    }

    fn evaluate(&mut self, expr: &Expr) -> Value {
	expr.accept(self)
    }

    fn is_truthy(&self, val: Value) -> bool {
	use Value::*;
	match val {
	    False => false,
	    True => true,
	    Nil => false,
	    Number(_) => true,
	    Strng(_) => true,
	}
    }

    fn is_equal(&self, a: Value, b: Value) -> bool {
	a == b
    }

    fn numeric(&self, operand: Value) -> f64 {
	match operand {
	    Value::Number(n) => n,
	    _ => panic!("operand must be a number")
	}
    }

}

impl stmt::Visitor<()> for Interpreter {
    fn visit_block(&mut self, stmts: &Vec<Stmt>) {
	let prev = self.environment.clone();
	let new = Rc::new(RefCell::new(Environment::from(&self.environment)));
	self.environment = new;
	for stmt in stmts.iter() {
	    self.execute(stmt);
	}
	self.environment = prev;
    }

    fn visit_print(&mut self, expr: &Expr) {
	let val = self.evaluate(expr);
	println!("{}", val);	
    }

    fn visit_expression(&mut self, expr: &Expr) {
	self.evaluate(expr);
    }

    fn visit_var(&mut self, name: &Token, initializer: &Expr) {
	let value = self.evaluate(initializer);
	self.environment.borrow_mut().define(&name.lexeme, value);	
    }
}

impl expr::Visitor<Value> for Interpreter {
    fn visit_assignment(&mut self, name: &Token, value: &Expr) -> Value {
	let val = self.evaluate(value);
	self.environment.borrow_mut().assign(name, val.clone());
	val
    }

    fn visit_literal(&self, val: &Value) -> Value {
	val.clone()
    }

    fn visit_grouping(&mut self, expression: &Expr) -> Value {
	self.evaluate(expression)
    }

    fn visit_binary(&mut self, left: &Expr, operator: &Token, right: &Expr) -> Value {
	let lt = self.evaluate(left);
	let rt = self.evaluate(right);
	
	use TokenType::*;
	// note: would be less verbose to implement
	// https://doc.rust-lang.org/std/ops/index.html
	// for the Value enum
	// then just apply them
	match operator.type_ {
	    Greater => { 
		Value::from(
		    self.numeric(lt) > self.numeric(rt)
		)
	    } ,
	    GreaterEqual => {
		Value::from(
		    self.numeric(lt) >= self.numeric(rt)
		)
	    },
	    Less => {
		Value::from(
		    self.numeric(lt) < self.numeric(rt)
		)
	    },
	    LessEqual => {
		Value::from(
		    self.numeric(lt) <= self.numeric(rt)
		)
	    },
	    Minus => {
		Value::from(
		    self.numeric(lt) - self.numeric(rt)
		)
	    },
	    Plus => {
		// handle string and number cases
		match (lt, rt) {
		    (Value::Strng(l), Value::Strng(r)) => Value::from(l + &r),
		    (Value::Number(l), Value::Number(r)) => Value::from(l + r),
		    (_, _) => panic!("operands to plus must both be numbers or both be strings"),
		}
	    },
	    Slash => {
		Value::from(
		    self.numeric(lt) / self.numeric(rt)
		)
	    },
	    Star => {
		Value::from(
		    self.numeric(lt) * self.numeric(rt)
		)
	    },
	    BangEqual => { Value::from(!self.is_equal(lt, rt)) },
	    EqualEqual => { Value::from(self.is_equal(lt, rt)) },
	    _ => Value::Nil
	}
    }

    fn visit_unary(&mut self, operator: &Token, expr: &Expr) -> Value {
	let right = self.evaluate(expr);
	match operator.type_ {
	    TokenType::Minus => {
		Value::from(-self.numeric(right))
	    },
	    TokenType::Bang => Value::from(!self.is_truthy(right)),
	    _ => panic!("unary operator should only be minus or bang")
	}
    }

    fn visit_variable(&self, name: &Token) -> Value {
	self.environment.borrow().get(name)
    }
}

File: jlox-java/jlox/Interpreter.java
package jlox;

import java.util.List;

class Interpreter implements Expr.Visitor<Object>, Stmt.Visitor<Void> {
  private Environment environment = new Environment();

  void interpret(List<Stmt> statements) {
    try {
      for (Stmt statement : statements) {
        execute(statement);
      }
    } catch (RuntimeError error) {
      Lox.runtimeError(error);
    }
  }

  private void execute(Stmt stmt) {
    stmt.accept(this);
  }

  @Override
  public Void visitBlockStmt(Stmt.Block stmt) {
    executeBlock(stmt.statements, this.environment);
    return null;
  }

  void executeBlock(List<Stmt> statements, Environment environment) {
    Environment previous = environment;
    try {
      this.environment = new Environment(previous);

      for (Stmt statement : statements) {
        execute(statement);
      }
    } finally {
      this.environment = previous;
    }
  }

  private Object evaluate(Expr expr) {
    return expr.accept(this);
  }

  @Override
  public Void visitExpressionStmt(Stmt.Expression stmt) {
    evaluate(stmt.expression);
    return null;
  }

  @Override
  public Void visitPrintStmt(Stmt.Print stmt) {
    Object value = evaluate(stmt.expression);
    System.out.println(stringify(value));
    return null;
  }

  @Override
  public Void visitVarStmt(Stmt.Var stmt) {
    Object value = Environment.undefined_var();
    if(stmt.initializer != null) {
      value = evaluate(stmt.initializer);
    }
    environment.define(stmt.name.lexeme, value);
    return null;
  }

  @Override 
  public Object visitAssignExpr(Expr.Assign expr) {
    Object value = evaluate(expr.value);

    environment.assign(expr.name, value);
    return value;
  }

  @Override
  public Object visitVariableExpr(Expr.Variable expr) {
    return environment.get(expr.name);
  }

  @Override
  public Object visitLiteralExpr(Expr.Literal expr) {
    return expr.value;
  }

  @Override
  public Object visitGroupingExpr(Expr.Grouping expr) {
    return evaluate(expr.expression);
  }

  @Override
  public Object visitUnaryExpr(Expr.Unary expr) {
    Object right = evaluate(expr.right);
    switch(expr.operator.type) {
      case MINUS:
        checkNumberOperand(expr.operator, right);
        return -(double)right;
      case BANG:
        return !isTruthy(right);
    }
    return null;
  }

  

  @Override
  public Object visitBinaryExpr(Expr.Binary expr) {
    Object left = evaluate(expr.left);
    Object right = evaluate(expr.right);

    switch(expr.operator.type) {
      case GREATER:
        checkNumberOperands(expr.operator, left, right);
        return (double)left > (double)right;
      case GREATER_EQUAL:
        checkNumberOperands(expr.operator, left, right);
        return (double)left >= (double)right;
      case LESS:
        checkNumberOperands(expr.operator, left, right);
        return (double)left < (double)right;
      case LESS_EQUAL:
        checkNumberOperands(expr.operator, left, right);
        return (double)left <= (double)right;
      case MINUS:
        checkNumberOperands(expr.operator, left, right);
        return (double)left - (double)right;
      case PLUS:
        if (left instanceof String || right instanceof String) {
          // added just for the challenge, but this is bad
          // instead, the above check should be &&
          // and this should be a normal (String) cast
          // instead of a call to stringify
          return stringify(left) + stringify(right);
        }

        if (left instanceof Double && right instanceof Double) {
          return (double)left + (double)right;
        }

        throw new RuntimeError(expr.operator,
            "Operands must be two numbers or two strings.");
      case SLASH:
        checkNumberOperands(expr.operator, left, right);
        checkNonZeroOperand(expr.operator, right);
        return (double)left / (double)right;
      case STAR:
        checkNumberOperands(expr.operator, left, right);
        return (double)left * (double)right;
      case BANG_EQUAL: return !isEqual(left, right);
      case EQUAL_EQUAL: return isEqual(left, right);
    }

    return null;
  }

  @Override
  public Object visitTernaryExpr(Expr.Ternary expr) {
    // ternary is short-circuiting, so only evaluate the first operand
    Object first = evaluate(expr.first);

    // ironically define ternary using ternary
    return isTruthy(first) ? evaluate(expr.second) : evaluate(expr.third);
  }


  private boolean isTruthy(Object object) {
    if (object == null) return false;
    if (object instanceof Boolean) return (boolean)object;
    return true;
  }

  private boolean isEqual(Object a, Object b) {
    if (a == null && b == null) return true;
    if (a == null) return false;
    return a.equals(b);
  }

  private String stringify(Object object) {
    if (object == null) return "nil";

    // Hack. Work around Java adding ".0" to integer-valued doubles.
    if (object instanceof Double) {
      String text = object.toString();
      if (text.endsWith(".0")) {
        text = text.substring(0, text.length() - 2);
      }
      return text;
    }

    return object.toString();
  }

  private void checkNumberOperand(Token operator, Object operand) {
    if (operand instanceof Double) return;
    throw new RuntimeError(operator, "Operand must be a number.");
  }

  private void checkNumberOperands(Token operator,
                                   Object left, Object right) {
    if (left instanceof Double && right instanceof Double) return;

    throw new RuntimeError(operator, "Operands must be numbers.");
  }

  private void checkNonZeroOperand(Token operator, Object right) {
    if((Double)right == 0) throw new RuntimeError(operator, "Division by zero.");
    return;
  }
}
